realm/realm-jni/src/io_realm_internal_TableQuery.cpp

```diff
@@ -970,58 +1114,149 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFindAll(
 
 // queryPtr would be owned and released by this function
 JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFindAllWithHandover
-  (JNIEnv *env, jobject, jlong bgSharedGroupPtr, jlong replicationPtr, jlong queryPtr, jlong start, jlong end, jlong limit)
+  (JNIEnv* env, jobject, jlong bgSharedGroupPtr, jlong replicationPtr, jlong queryPtr, jlong start, jlong end, jlong limit)
   {
       TR_ENTER()
       try {
           std::unique_ptr<Query> query = getHandoverQuery(bgSharedGroupPtr, replicationPtr, queryPtr);
-          TableRef table = query->get_table();
-          if (!QUERY_VALID(env, query.get()) ||
-              !ROW_INDEXES_VALID(env, table.get(), start, end, limit)) {
-              return 0;
-          }
-
-          // run the query
-          TableView tableView(query->find_all(S(start), S(end), S(limit)));
-
-          // handover the result
-          std::unique_ptr<SharedGroup::Handover<TableView>> handover = SG(
-                  bgSharedGroupPtr)->export_for_handover(tableView, MutableSourcePayload::Move);
-          return reinterpret_cast<jlong>(handover.release());
+          return findAllWithHandover(env, bgSharedGroupPtr, std::move(query), start, end, limit);
       } CATCH_STD()
       return 0;
   }
 
+
+
+// Should match the values in Java ArgumentsHolder class
+enum query_type {QUERY_TYPE_FIND_ALL = 0, QUERY_TYPE_DISTINCT = 4, QUERY_TYPE_FIND_ALL_SORTED = 1, QUERY_TYPE_FIND_ALL_MULTI_SORTED = 2};
+
+// batch update of async queries
+JNIEXPORT jlongArray JNICALL Java_io_realm_internal_TableQuery_nativeBatchUpdateQueries
+        (JNIEnv *env, jobject, jlong bgSharedGroupPtr, jlong replicationPtr,
+         jlongArray  handover_queries_array /*list of handover queries*/,
+         jobjectArray  query_param_matrix /*type & params of the query to be updated*/,
+         jobjectArray  multi_sorted_indices_matrix,
+         jobjectArray  multi_sorted_order_matrix)
+{
+    TR_ENTER()
+    try {
+        JniLongArray handover_queries_pointer_array(env, handover_queries_array);
+
+        const size_t number_of_queries = env->GetArrayLength(query_param_matrix);
+
+        jlong* exported_handover_tableview_array = new jlong[number_of_queries];
+        if (exported_handover_tableview_array == NULL) {
```

Comment:
@kneth says you are an old C-programmer ;) In any case, all these checks can be removed since new will throw an exception instead of returning NULL: http://www.cplusplus.com/reference/new/operator%20new/